src/QsCompiler/CompilerOptimizations/Evaluation.fs

```diff
@@ -13,170 +13,181 @@ open Microsoft.Quantum.QsCompiler.SyntaxTree
-/// I use the Result type to match the common design pattern of railway-oreiented programming.
-type private FunctionState = Result<Constants<TypedExpression>, FunctionInterrupt>
+/// A shorthand for the specific Imperative type used by several functions in this file
+type private Imp<'t> = Imperative<EvalState, 't, FunctionInterrupt>
+
+/// Represents a computation that decreases the remaining statements counter by 1.
+/// Yields an OutOfStatements interrupt if this decreases the remaining statements below 0.
+let private incrementState: Imp<Unit> = imperative {
+    let! vars, counter = getState
+    if counter < 1 then yield OutOfStatements
+    do! putState (vars, counter - 1)
+}
+
+/// Represents a computation that sets the given variables to the given values
+let private setVars callables entry: Imp<Unit> = imperative {
+    let! vars, counter = getState
+    do! putState (defineVarTuple (isLiteral callables) vars entry, counter)
+}
 
 
 /// Evaluates functions by stepping through their code
-type internal FunctionEvaluator(callables: Callables, maxRecursiveDepth: int) =
+type internal FunctionEvaluator(callables: Callables) =
 
-    /// Transforms a BoolLiteral into the corresponding bool
-    let castToBool f x =
+    /// Casts a BoolLiteral to the corresponding bool
+    let castToBool x: Imp<bool> = imperative {
         match x.Expression with
-        | BoolLiteral b -> Ok (f b)
-        | _ -> "Not a BoolLiteral: " + (printExpr x.Expression) |> CouldNotEvaluate |> Error
-
-    /// The callback for the subroutine that evaluates and simplifies an expression
-    member internal this.evaluateExpression constants expr =
-        ExpressionEvaluator(callables, constants, maxRecursiveDepth - 1).Transform expr
+        | BoolLiteral b -> return b
+        | _ -> yield CouldNotEvaluate ("Not a BoolLiteral: " + (printExpr x.Expression))
+    }
+
+    /// Evaluates and simplifies a single Q# expression
+    member internal __.evaluateExpression expr: Imp<TypedExpression> = imperative {
+        let! vars, counter = getState
+        let result = ExpressionEvaluator(callables, vars, counter / 2).Transform expr
+        if isLiteral callables result then return result
+        else yield CouldNotEvaluate ("Not a literal: " + (printExpr result.Expression))
+    }
 
     /// Evaluates a single Q# statement
-    member private this.evaluateStatement (constants: Constants<TypedExpression>) (statement: QsStatement): FunctionState =
-        let eval constantsRef newScope scope = result {
-            let! s = this.evaluateScope !constantsRef newScope scope
-            constantsRef := s }
+    member private this.evaluateStatement (statement: QsStatement): Imp<Unit> = imperative {
+        do! incrementState
 
         match statement.Statement with
-        | QsExpressionStatement expr ->
-            this.evaluateExpression constants expr |> ignore; constants |> Ok
+        | QsExpressionStatement _ ->
+            ()
```

Comment:
Where are expression statements evaluated now / why did this change?

Reply:
Since we're inside of a function, expression statements will never have side effects (as they can't mutate state), so we can skip evaluating them. The previous code was effectively doing the same - see the call to ignore.

Reply:
Right - a comment here would be helpful.

Reply:
Fixed in latest commit